### ISSUE TITLE
[$500] [Live Markdown] Copying and pasting from google docs comes over as bold

### ISSUE THREAD

**Issue by thienlnam** [Expensify/App#41109]
1. In a google doc, highlight and copy some regular text
2. Paste it into the composer 

Actual
![image](https://github.com/Expensify/App/assets/30609178/2e97eacc-900e-4d7f-b794-c17650aba3da)
![image](https://github.com/Expensify/App/assets/30609178/c529c4e5-5fdf-4afc-afcd-a734d287fbd9)

Expectation
It gets pasted as regular text

cc @tomekzaw 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0181c168843a4dc29c</li>
        <li>Upwork Job ID: 1783979270880542720</li>
        <li>Last Price Increase: 2024-05-20</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 102487686</li>
<li>ShridharGoel | Contributor | 102487688</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @abekkala</details>

**Comment by climax0916** [comment]
@thienlnam Is it okay to prevent all formatting when paste text?

**Comment by thienlnam** [comment]
We'd like to retain the formatting that it was copied with, instead of preventing all formatting at all

**Comment by jainilparikh** [comment]
@thienlnam , isn't this the expected behaviour ? If the text was bold on Google docs, we should send it as bold in our chat too. 

The composer is showing * text * since it's allowing the end user to choose whether they want to continue with the formatting that their copied text used on Google docs or whether they want to change it. In-case they want to change it, they can use markdown (which the composer supports) to change the formatting.

**Comment by skyweb331** [comment]
@thienlnam 
When I copy  "**Test**123" from google doc, it is translated into HTML like this.
```
<html>
<body>
<!--StartFragment--><meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ef0d09b9-7fff-2c97-04c0-4fe5952c2707"><span style="font-size:11pt;font-family:Arial,sans-serif;color:#000000;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">TEST</span><span style="font-size:11pt;font-family:Arial,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"> 12</span></b><!--EndFragment-->
</body>
</html>
```
Too much complicated, but simply it is 
`<b><span style="font-weight: 700">Test</span><span style="font-weight:400">123</span></b>`

But, `ExpensiMark` does not detect the font-weight, it is detecting `<b>` tag, that's why pasted text is parsed into `*Test 123*` in our editor.

As I mentioned in https://github.com/Expensify/App/issues/41110#issuecomment-2080034830, this could not be done by regex. Regex has its limitation and could not parse all these complex html.  ( I finally find the exact example that regex could not fit for every cases )

So my proposal is to change the HTML parsing logic like using `xml2js` or `fast-xml-parser` (https://github.com/Expensify/App/issues/40571#issuecomment-2070080632)

This way, we can completely resolve the root problem.

(cc @jjcoffee )

**Comment by tomekzaw** [comment]
@skyweb331 We could just create DOM subtree for the pasted HTML and then get pure text using the following approach: https://stackoverflow.com/a/6743966/23325954

**Comment by skyweb331** [comment]
@tomekzaw 
Yes. If we use DOM, solution is really simple. I also recommended to use DOM, but on mobile device, DOM parsing is available but slow...That's why `ExpensiMark` is using regex. `xml2js` and `fast-xml-parser` is not using DOM and I recommend it.
( I had discussion with @ikevin127 for this. https://expensify.slack.com/archives/C01GTK53T8Q/p1713721057329569 )

**Comment by thienlnam** [comment]
Could you put a post in #expensify-open-source about the problem and the solution solved by using another parser so we can get some more eyes on it?

**Comment by skyweb331** [comment]
@thienlnam https://expensify.slack.com/archives/C01GTK53T8Q/p1713721057329569
I already posted this and discussed with Kevin Bader.
We have all parsing Markdown to HTML and vice-versa in One file `ExpensiMark` and using regex for everything. It is not well-structured and hard to maintain...

**Comment by abekkala** [comment]
@alitoshmatov can you review the proposal above please?

**Comment by alitoshmatov** [comment]
Thank you for your proposal @ShridharGoel while your RCA is correct I don't think you solution is optimal one

**Comment by skyweb331** [comment]
To prevent future issues and improve the logic easily, `ExpensiMark` should change its codebase.

Google doc is using `font-weight` attributes to make it bold, we can implement the logic above, but  what if Slack or others' are using different tags or different attributes? Maybe, the regex pattern will be more complicated and at one point, no longer can be updated...

Solution is to clean up current `ExpensiMark` structure...

**Comment by thienlnam** [comment]
Increasing the bounty to get some more eyes

**Comment by skyweb331** [comment]
@thienlnam This is not related to bounty problem... It requires `ExpensiMark` rebuild.

**Comment by thienlnam** [comment]
I understand your concerns about parsing HTML with regex and your suggestion to rebuild ExpensiMark. However, without a detailed proposal, it's challenging to evaluate the necessity and feasibility of such a significant change.

If you believe a rebuild is the best solution, could you please provide a formal proposal? This should include things like:
- Specific issues and limitations with the current approach.
- The benefits of transitioning to a new parser.
- Detailed steps and an implementation plan for the rebuild.
- An analysis of the potential challenges and how they can be addressed.


@alitoshmatov Could you respond to this comment? https://github.com/Expensify/App/issues/41109#issuecomment-2101316189

**Comment by alitoshmatov** [comment]
> Can you tell what can be improved?

@ShridharGoel To be honest I am not sure. I am just worried if this will cover all cases, I mean based on your solution if I copy the bold tag without any styles will it be parsed as bold. What would be our requirement for any element to be parsed as bold

**Comment by abekkala** [comment]
@thienlnam do you have any input on the proposal and the comments after?

**Comment by alitoshmatov** [comment]
Okay let's go with @ShridharGoel 's [proposal](https://github.com/Expensify/App/issues/41109#issuecomment-2095057640). Which parses text in one of these two scenarios:
1. The copied tag is `<b>` and doesn't have any `font-weight` attributes
2. The copied tag is `<b>` and has `font-weight` attribute with values of either `bold` or `700`

C+ reviewed 🎀 👀 🎀

**Comment by marcochavezf** [comment]
Sounds good, thanks @alitoshmatov for the review. Assigning @ShridharGoel 🚀

**Comment by marcochavezf** [comment]
Update: PR in review

**Comment by abekkala** [comment]
Not overdue - PR is in review

**Comment by marcochavezf** [comment]
@ShridharGoel in working on the PR

**Comment by zfurtak** [comment]
Hi there 😊 So far it's been established in live markdown that text should be pasted without formatting even if it has it in google docs eg. The only case is when text is pasted explicitly with `*   *`. 
After this fix: **bold** is pasted with formatting but for example _italic_ and underline are not.

@thienlnam what is the approach we want to follow?

**Comment by tomekzaw** [comment]
I'd recommend pasting text without formatting, i.e. if someone copies "Hello **world**" from Google Docs where **world** is bold, it should be pasted as "Hello world" so that the resulting message as a JS string is `Hello world`.

Obviously, if the copied text is "Hello \*world\*", it should be previewed as "Hello \***world**\*" and the resulting JS string would be `Hello *world*`.

**Comment by abekkala** [comment]
@thienlnam @marcochavezf 

> @thienlnam what approach do we want to take? Do we want to continue pasting without formatting or do we want to change approach completely?

What are your thoughts on this? 

Along with the comment from @tomekzaw 

> I'd recommend pasting text without formatting, i.e. if someone copies "Hello world" from Google Docs where world is bold, it should be pasted as "Hello world" so that the resulting message as a JS string is Hello world.
> 
> Obviously, if the copied text is "Hello *world*", it should be previewed as "Hello *world*" and the resulting JS string would be Hello *world*.

**Comment by marcochavezf** [comment]
> @thienlnam @marcochavezf
> 
> > @thienlnam what approach do we want to take? Do we want to continue pasting without formatting or do we want to change approach completely?
> 
> What are your thoughts on this?
> 
> Along with the comment from @tomekzaw
> 
> > I'd recommend pasting text without formatting, i.e. if someone copies "Hello world" from Google Docs where world is bold, it should be pasted as "Hello world" so that the resulting message as a JS string is Hello world.
> > Obviously, if the copied text is "Hello _world_", it should be previewed as "Hello _world_" and the resulting JS string would be Hello _world_.

Brought this to Slack for consensus:  https://expensify.slack.com/archives/C01GTK53T8Q/p1718407088175969

**Comment by marcochavezf** [comment]
https://github.com/Expensify/expensify-common/pull/710 in review, @alitoshmatov can you review it when you can? :)

**Comment by alitoshmatov** [comment]
Working on it

**Comment by zfurtak** [comment]
@marcochavezf 
based on what @tomekzaw explained in the [slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1718407088175969) and here above in a comment, this proposal doesn't work like it's expected.
As it was agreed, **bold text** copied from google docs should be pasted also as normal font, therefore this Proposal and PR should be changed.
The only case when the text is pasted as bold is when it has asterisks explicitly written `*proper bold text*`

**Comment by alitoshmatov** [comment]
Thank you @zfurtak for your feedback. Our aim here on selected proposal was to keep formatting of the pasted text same as it is in google docs. I see we might have overlooked some issues here

**Comment by alitoshmatov** [comment]
@ShridharGoel I have noticed this cases
When we paste these text from google docs the html is following
1.  Hello **world** -> ```<b style="font-weight: normal"><span style="font-weight: 400">Hello </span><span style="font-weight: 700">world</span></b>```
2. **Hello** -> `<b style="font-weight: normal"><span style="font-weight: 700">Hello </span></b>`
3. Hello ->  `<b style="font-weight: normal"><span style="font-weight: 400">Hello </span></b>`
4. _Hello_ -> `<b style="font-weight: normal"><span style="font-weight: 400; font-style: italic">Hello </span></b>`

In microsoft word
1. Hello **world**, _new order_  -> ```<span style="font-style: normal; font-weight: 400;"><span>Hello </span></span><span style="font-weight: bold;"><span>world, </span></span><span style="font-weight: normal; font-style: italic;"><span >new order</span></span>```

**Comment by alitoshmatov** [comment]
@ShridharGoel I think we should revisit your solution and cover all the cases

**Comment by marcochavezf** [comment]
👋🏽 @ShridharGoel let us know when the PR is ready for review with the updated changes

**Comment by zfurtak** [comment]
Hi @ShridharGoel 
I have one more thing 😊 
Have you tried if it works on different platforms? I tried on IOS and it didn't work for me. Could you take a look at it? Because we want features to work the same across platforms.

**Comment by zfurtak** [comment]
Okay then 👍

**Comment by jasperhuangg** [comment]
@ShridharGoel I think it's fine to only handle the bold case for this particular issue. @alitoshmatov thoughts?

**Comment by marcochavezf** [comment]
I agree to handle only bold here cc @alitoshmatov

**Comment by marcochavezf** [comment]
Waiting on @alitoshmatov's review https://github.com/Expensify/expensify-common/pull/710#pullrequestreview-2157166019

**Comment by marcochavezf** [comment]
https://github.com/Expensify/expensify-common/pull/710 merged

**Comment by marcochavezf** [comment]
@ShridharGoel @alitoshmatov given the PR for expensify-common was merged, I suspect we'd need another PR to bump the package in App, right?

**Comment by alitoshmatov** [comment]
We can start regression period, other similar PR was quicker then ours and bumped `expensify-common` to the latest version which also includes our changes. So we can close our PR, and start regression period.

**Comment by abekkala** [comment]
@thienlnam @marcochavezf I noticed that the PR for this issue was closed without merging based on another PR fixing this issue sooner. 

Can you confirm if payments are still payable to both contributors here?

**Comment by marcochavezf** [comment]
@abekkala yup it's still payable, since the PR that was closed was to bump the version in App, but the main work was done in another PR

**Comment by abekkala** [comment]
### PAYMENT SUMMARY: 

- Fix: @ShridharGoel [$500] the original offer expired in Upwork. new one sent! [Offer](https://www.upwork.com/nx/wm/offer/103328711)
- PR Review: @alitoshmatov [$500] [Offer](https://www.upwork.com/nx/wm/workroom/37412052/overview)

**Comment by abekkala** [comment]
@alitoshmatov payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
@ShridharGoel payment sent and contract ended - thank you! 🎉
